### PR TITLE
Addressed rotation problems in RangeWidget under RTL languages. Fix #1983

### LINKED ISSUE #1983: Problems with Range Widgets with RTL languages
#### Software and hardware versions 
Collect v1.13.2

#### Problem description
Problems with Range Widgets with RTL languages

#### Steps to reproduce the problem
1. User uses RTL languages
2. User opens Range widget
3. Horizontal widget shows incorrect value on Android 4.4 (Vertical widget works as expected)

| Integer | Decimal  | 
| ---- | ----|
| screenshot_2018-03-07-13-19-56 | screenshot_2018-03-07-13-20-29 |

4. Vertical widget shows incorrect value on Android 5.1,  6.0, 7.0, 8.0

| Integer | Decimal  |
|----|----|
| screenshot_20180307-164121 | screenshot_20180307-183643|

5. Vertical widget is not centred on Android 4.2 and values are incorrect
6. Min and max values are overlapped in horizontal widget on Android 4.2

|  |  |  | | 
| ---- | ---- | ---- | ---- |
| screenshot_2018-03-07-13-37-41 | screenshot_2018-03-07-13-38-02 | screenshot_2018-03-07-13-37-56 |screenshot_2018-03-07-13-37-50|

#### Expected behavior
* Range Widget always shows correct values

### PR DESCRIPTION
**Pull request by huangyz0918**
Closes #1983  

#### What has been done to verify that this works as intended?
I found a difference in native Android views. With RTL languages environment, Android would draw views in different directions with different API levels.(API level > 17 and API level 16). Besides, when we are under a RTL language circumstance, the routations may be totally different from LTR languages. So we need to check the Android version, the System Language and reverse the views if necessary.

Here are some tests records:

__Form and Form list:__

|   Android 4.1.1 (API level 16)           |  Android 4.4 (API level 19)      |
| ---- | :----: |
|||
|||

|   Android 5.0 (API level 21)           |  Android 6.0 (API level 23)      |
| ---- | :----: |
|||
|||

|   Android 7.0 (API level 24)           |  Android 8.0 (API level 26)      |
| ---- | :----: |
|||
|||

#### Why is this the best possible solution? Were any other approaches considered?
Because these problems were caused by Android System, difference API would cause different behaviors. So I think this is best way to address this problem.

#### Are there any risks to merging this code? If so, what are they?
Na
#### Do we need any specific form for testing your changes? If so, please attach one.
No, this problem only relate to RangeWidget.

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/widgets/RangeWidget.java [hunk 1]
```diff
@@ -203,6 +217,17 @@ public boolean onTouch(View v, MotionEvent event) {
         });
     }
 
+    // check if it's RTL languages
+    public static boolean isRTL() {
+        return isRTL(Locale.getDefault());
+    }
+
+    public static boolean isRTL(Locale locale) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/widgets/RangeWidget.java]: Both those methods already exist in `QuestionWidget` so please use them.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/widgets/RangeWidget.java (hunk 1 above)
Comment (huangyz0918) [reply collect_app/src/main/java/org/odk/collect/android/widgets/RangeWidget.java]: Thanks for remaining !
I will update my submits ;)
Ps, I think PR#2048 also need review if you have spare time.
Have a good day !

**Comment by huangyz0918** [comment]
@grzesiek2010 I use ` ./gradlew collect_app:transformClassesWithDesugarForDebug --stacktrace` to check my code in local device, it tested ok.
I haven't changed anything, but CircleCI PMD shows an error in `transformClassesWithDesugarForDebug`, so I guess it may need a Gradle refresh in a second CI test. Maybe it's just a environment issue.
Am I right? 

|   My test result          |  Circle Ci test result      |
| ---- | :----: |
|||

**Comment by huangyz0918** [comment]
@opendatakit-bot label "needs review"

File: collect_app/src/main/java/org/odk/collect/android/widgets/RangeWidget.java [hunk 2]
```diff
@@ -179,6 +179,19 @@ private void setUpSeekBar() {
         seekBar.setMax(elementCount);
         seekBar.setProgress(progress);
         seekBar.setOnSeekBarChangeListener(this);
+        // fix Problems with Range Widgets with RTL languages #1983.
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/widgets/RangeWidget.java]: This comment is not needed, you are in RangeWidget class so it's clear you solve the problem wit RangeWidget and `isRTL()` method name is also descriptive enough.

**Review by grzesiek2010 — CHANGES_REQUESTED** [review]

**Comment by grzesiek2010** [comment]
I tried `./gradlew lint checkstyle findbugs pmd` on your branch and everything is ok. Could you click on `Rerun job` button to try again?

**Comment by huangyz0918** [comment]
@grzesiek2010 ok, I'm trying 
thanks

**Comment by grzesiek2010** [comment]
@opendatakit-bot label "needs testing"

**Review by grzesiek2010 — APPROVED** [review]

**Comment by kkrawczyk123** [comment]
### Bugs have been found!
@huangyz0918 @lognaturel 
Verified on Androids: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0 and 8.0
1. Axis is not reversed on Android 4.1. Screenshot attached:
screenshot_2018-03-27-13-19-01
2. Issues from point 6 from @mmarciniak90 [comment](https://github.com/opendatakit/collect/issues/1983#issue-303085771) connected to Android 4.2 have not been fixed at all.

**Comment by huangyz0918** [comment]
The Axis in 4.1 looks good, it's value is correct already, so we don't have to reverse. @kkrawczyk123 I think it's not a bug, it had been addressed and we can get a correct value now.

As for the question related to overlapped values in 4.2, I'm sorry I just ignored this problem before...
I can fix this in a few minutes, thanks for reminding :)

**Comment by huangyz0918** [comment]
okey, The overlapped values disappeared! Here is the screenshot:
 

And why don't we have to reverse the Axis in API 16? Because Android only supports the RTL language in devices with API level higher than 17. In device with API 16, this issue won't happen, so I didn't change anything in this API, it gets a correct value already.

btw, I have checked the code before pushing a commit, it tested ok. I have viewed the CI log, but still cannot figure out why it failed in this PR. I don't have a write permission, so I cannot start the `return job` in CI @grzesiek2010 , thanks anyway.

**Comment by mmarciniak90** [comment]
@huangyz0918 Thanks for so quick fix! Unfortunately, the vertical range widget is not centered on Android 4.2

| | |
| ---| --- |
|screenshot_2018-03-27-15-18-20 | screenshot_2018-03-27-15-18-39 |

**Comment by huangyz0918** [comment]
@mmarciniak90 I cannot reproduce this bug, 4.2 looks like this in my device:
 

Are you sure it's a common issue ? If not, I cannot know what makes this happen. But in my view, I think this may also caused by layouts, I can create a special layout for 4.2 version immediately, would you mind help me test it again ? thanks :)

**Comment by mmarciniak90** [comment]
@huangyz0918 I have only one device with Android 4.2. Of course, I will test it. Let me know when it will be ready.

**Comment by huangyz0918** [comment]
@mmarciniak90 Thanks for understanding, I have finished what I can do to address your issue and it's ready for testing.

**Comment by mmarciniak90** [comment]
@huangyz0918 Unfortunately, fix does not work. But I talked with @lognaturel and we decided that it is not worth to spend more time on this. We can leave it not centered on Android 4.2.
So, please remove last change and we will merge it.

**Comment by huangyz0918** [comment]
@mmarciniak90  I don't know what happened in your device, so it's difficult to help you, I'm sorry.
Thanks for a further testing 👍

**Comment by mmarciniak90** [comment]
**Tested with success!**

Verified on Android: quick testing on 4.2, 6.0

**Known Issues:**
* Vertical range widget could be not centered on Android 4.2
* Axis is not reversed on API 16 because the RTL language is supported in devices with API > 16.

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Review by lognaturel — APPROVED** [review]

**Comment by lognaturel** [comment]
Thank you @huangyz0918, @mmarciniak90, @kkrawczyk123!

**Pull request merged**